src/util/createServerRootMixin.js

```diff
@@ -0,0 +1,255 @@
+import algoliaHelper from 'algoliasearch-helper';
+const { SearchResults, SearchParameters } = algoliaHelper;
+import { warn } from './warn';
+
+function walkIndex(indexWidget, visit) {
+  visit(indexWidget);
+
+  return indexWidget.getWidgets().forEach(widget => {
+    if (widget.$$type !== 'ais.index') return;
+    visit(widget);
+    walkIndex(widget, visit);
+  });
+}
+
+function renderToString(app) {
+  return new Promise((resolve, reject) =>
+    _renderToString(app, (err, res) => {
+      if (err) reject(err);
+      resolve(res);
+    })
+  );
+}
+
+function searchOnlyWithDerivedHelpers(helper) {
+  return new Promise((resolve, reject) => {
+    helper.searchOnlyWithDerivedHelpers();
+
+    // we assume all derived helpers resolve at least in the same tick
+    helper.derivedHelpers[0].on('result', e => {
+      resolve(e);
+    });
+
+    helper.derivedHelpers.forEach(derivedHelper =>
+      derivedHelper.on('error', e => {
+        reject(e);
+      })
+    );
+  });
+}
+
+function augmentInstantSearch(instantSearchOptions, searchClient, indexName) {
+  /* eslint-disable no-param-reassign */
+
+  const helper = algoliaHelper(searchClient, indexName);
+  const search = instantsearch(instantSearchOptions);
+
+  let resultsState;
+
+  /**
+   * main API for SSR, called in serverPrefetch of a root component which contains instantsearch
+   * @param {object} componentInstance the calling component's `this`
+   * @returns {Promise} result of the search, to save for .hydrate
+   */
+  search.findResultsState = function(componentInstance) {
+    let app;
+    return Promise.resolve()
+      .then(() => {
+        const options = {
+          serverPrefetch: undefined,
+          fetch: undefined,
+          _base: undefined,
+          name: 'ais-ssr-root-component',
+        };
+
+        const extended = componentInstance.$vnode
+          ? componentInstance.$vnode.componentOptions.Ctor.extend(options)
+          : Object.assign({}, componentInstance.$options, options);
+
+        app = new Vue(extended);
+
+        app.$options.serverPrefetch = [];
+
+        app.instantsearch.helper = helper;
+        app.instantsearch.mainHelper = helper;
+
+        app.instantsearch.mainIndex.init({
+          instantSearchInstance: app.instantsearch,
+          parent: null,
+          uiState: app.instantsearch._initialUiState,
+        });
+      })
+      .then(() => renderToString(app))
+      .then(() => searchOnlyWithDerivedHelpers(helper))
+      .then(() => {
+        const results = {};
+        walkIndex(app.instantsearch.mainIndex, widget => {
+          results[widget.getIndexId()] = widget.getResults();
+        });
+
+        search.hydrate(results);
+
+        resultsState = Object.keys(results)
+          .map(indexId => {
+            const { _state, _rawResults } = results[indexId];
+            return [
+              indexId,
+              {
+                _state: JSON.parse(JSON.stringify(_state)),
```

Comment:
Can you explain why we stringify and parse it again?

Reply:
_state is a class, so we want to get all the keys, and only those which aren't functions. toJson is the simplest way to trigger that. In principle we could change this to Object.keys.reduce instead

Reply:
Ah got it. Can you add this comment in the code?